### PR TITLE
Set finalizers on Korifi resources in a webhook

### PR DESCRIPTION
**Pull request by danail-branekov**
## Is there a related GitHub Issue?
#2367

## What is this change about?
By setting the finalizer in a webhook we ensure that the finalizer is
installed as part of the object creation thus avoiding potential race
conditions. Previously, finalizers used to be installed in the
resources' controllers that could happen before the resource has been
deleted.

We have introduced 2 webhooks:
* one for the "core" Korifi resources, such as CFOrg, CFSpace, CFApp,
  etc.
* one for the kpack-image-builder related resources - BuildWorkload and
  korifi kpack builds

The reason to have them split in this manner is that the
kpack-image-builder is an optional component, if disabled, we should not
be installing our finalizers as custom BuildWorkload controllers may not
need them.

Now that we are setting finalizers in webhooks, we had to enabled
webhooks for all integration tests that tested finalization. This
results in massive webhook wiring as all the webhooks listed on the
webhooks manifest.yaml has to be wired. Not doing that causes envtest
complaining that no implementation has been registered per webhook
configuration.

Installing webhooks (both validating and mutating) means that we need to
ensure that tests are creating valid objects, which caused another chunk
of changes in tests' setup.

## Does this PR introduce a breaking change?
No

## Acceptance Steps
Not a functional change, everything should work as before

## Tag your pair, your PM, and/or team
@gcapizzi

### CODE REVIEW

File: controllers/controllers/networking/cfroute_controller.go [hunk 1]
```diff
@@ -473,5 +463,5 @@ func generateServiceName(destination *korifiv1alpha1.Destination) string {
 }
 
 func buildFQDN(cfRoute *korifiv1alpha1.CFRoute, cfDomain *korifiv1alpha1.CFDomain) string {
-	return strings.ToLower(fmt.Sprintf("%s.%s", cfRoute.Spec.Host, cfDomain.Spec.Name))
+	return fmt.Sprintf("%s.%s", cfRoute.Spec.Host, cfDomain.Spec.Name)
```
Comment (kieron-dev) [inline controllers/controllers/networking/cfroute_controller.go:466]: Why are we removing the strings.ToLower() here?

File: controllers/controllers/networking/cfroute_controller_test.go [hunk 2]
```diff
@@ -144,36 +157,19 @@ var _ = Describe("CFRouteReconciler Integration Tests", func() {
 		}).Should(Succeed())
 	})
 
-	When("the route Host contains upper case characters", func() {
```
Comment (kieron-dev) [inline controllers/controllers/networking/cfroute_controller_test.go:147]: Is it safe to be removing this test?

File: controllers/controllers/workloads/cfapp_controller_test.go [hunk 3]
```diff
@@ -325,71 +340,6 @@ var _ = Describe("CFAppReconciler Integration Tests", func() {
 			})
 		})
 
-		When("routes exist and CFProcesses do not exist for the app", func() {
```
Comment (kieron-dev) [inline controllers/controllers/workloads/cfapp_controller_test.go:328]: Why is this being removed?

File: helm/korifi/kpack-image-builder/manifests.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    clientConfig:
+      service:
+        name: korifi-controllers-webhook-service
+        namespace: '{{ .Release.Namespace }}'
+        path: /mutate-korifi-cloudfoundry-org-v1alpha1-kpack-image-builder-finalizer
+    failurePolicy: Fail
+    name: mcf-kib-finalizer.korifi.cloudfoundry.org
+    rules:
+      - apiGroups:
+          - korifi.cloudfoundry.org
```
Comment (kieron-dev) [inline helm/korifi/kpack-image-builder/manifests.yaml:21]: It's a bit weird how these resources have been mixed here. It looks like we could be operating on korifi.cloudfoundry.org/build and kpack.io/buildworkload, even though those don't exist.

I guess it's just a kubebuilder generation thing, and not massively important.

File: kpack-image-builder/controllers/kpack_build_controller.go [hunk 5]
```diff
@@ -68,41 +69,43 @@ func (c KpackBuildController) SetupWithManager(mgr manager.Manager) *ctrl.Builde
...
 	log := c.log.WithValues("namespace", kpackBuild.Namespace, "name", kpackBuild.Name, "deletionTimestamp", kpackBuild.DeletionTimestamp)
+	log.Info("in finalizer")
 
 	if !kpackBuild.GetDeletionTimestamp().IsZero() {
-		if !controllerutil.ContainsFinalizer(kpackBuild, kpackBuildFinalizer) {
+		if !controllerutil.ContainsFinalizer(kpackBuild, KpackBuildFinalizer) {
 			return ctrl.Result{}, nil
 		}
 
+		fmt.Println("in finalizer and deleted")
```
Comment (kieron-dev) [inline kpack-image-builder/controllers/kpack_build_controller.go:79]: Remove fmt.Println()

File: kpack-image-builder/controllers/kpack_build_controller.go [hunk 6]
```diff
@@ -68,41 +69,43 @@ func (c KpackBuildController) SetupWithManager(mgr manager.Manager) *ctrl.Builde
...
 		if kpackBuild.Status.LatestImage != "" {
 			tagsToDelete := []string{}
+			fmt.Printf("kpackBuild.Spec.Tags = %+v\n", kpackBuild.Spec.Tags)
```
Comment (kieron-dev) [inline kpack-image-builder/controllers/kpack_build_controller.go:82]: remove

File: kpack-image-builder/controllers/kpack_build_controller.go [hunk 7]
```diff
@@ -68,41 +69,43 @@ func (c KpackBuildController) SetupWithManager(mgr manager.Manager) *ctrl.Builde
...
 			for _, t := range kpackBuild.Spec.Tags {
 				parts := strings.Split(t, ":")
 				if len(parts) == 2 {
 					tagsToDelete = append(tagsToDelete, parts[1])
 				}
 			}
 
+			fmt.Println("about to delete image")
```
Comment (kieron-dev) [inline kpack-image-builder/controllers/kpack_build_controller.go:90]: remove

File: kpack-image-builder/controllers/kpack_build_controller.go [hunk 8]
```diff
@@ -68,41 +69,43 @@ func (c KpackBuildController) SetupWithManager(mgr manager.Manager) *ctrl.Builde
...
 			err := c.imageDeleter.Delete(ctx, image.Creds{
 				Namespace:          kpackBuild.Namespace,
 				ServiceAccountName: c.registryServiceAccount,
 			}, kpackBuild.Status.LatestImage, tagsToDelete...)
 			if err != nil {
+				fmt.Println("falied to delete image")
```
Comment (kieron-dev) [inline kpack-image-builder/controllers/kpack_build_controller.go:96]: remove

File: kpack-image-builder/controllers/kpack_build_controller.go [hunk 9]
```diff
@@ -68,41 +69,43 @@ func (c KpackBuildController) SetupWithManager(mgr manager.Manager) *ctrl.Builde
...
 				log.Info("failed to delete droplet image", "reason", err)
 			}
 		}
 
-		if controllerutil.RemoveFinalizer(kpackBuild, kpackBuildFinalizer) {
+		fmt.Println("about to remove finalizer")
```
Comment (kieron-dev) [inline kpack-image-builder/controllers/kpack_build_controller.go:101]: remove

File: kpack-image-builder/controllers/kpack_build_controller_test.go [hunk 10]
```diff
@@ -107,27 +107,4 @@ var _ = Describe("KpackBuildReconciler", func() {
 			})
 		})
 	})
-
-	Context("non-Korifi builds", func() {
```
Comment (kieron-dev) [inline kpack-image-builder/controllers/kpack_build_controller_test.go:111]: Is this safe to remove?

**Review by kieron-dev — COMMENTED** [review]
Some fmt.Prints to remove, and also wondering why a couple of tests are being removed.

File: controllers/api/v1alpha1/cfprocess_webhook_test.go [hunk 11]
```diff
@@ -32,7 +32,7 @@ var _ = Describe("CFProcessMutatingWebhook", func() {
 		cfProcess = &korifiv1alpha1.CFProcess{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      cfProcessGUID,
-				Namespace: namespace,
+				Namespace: rootNamespace,
```
Comment (matt-royal) [inline controllers/api/v1alpha1/cfprocess_webhook_test.go:35]: CFProcesses won't exist in the root namespace in practice, so I'd undo this change.

**Review by matt-royal — DISMISSED** [review]
I'm concerned about the use of the root namespace for resources that will never be placed there. Otherwise, this looks good.

File: controllers/controllers/workloads/cfapp_controller_test.go (hunk 3 above)
Comment (kieron-dev) [reply controllers/controllers/workloads/cfapp_controller_test.go:328]: Ok - this is a duplicate test and missing process creation has nothing to do with routes.

File: controllers/controllers/networking/cfroute_controller_test.go (hunk 2 above)
Comment (kieron-dev) [reply controllers/controllers/networking/cfroute_controller_test.go:147]: OK - validation stops any uppercase characters being used, so this was redundant.

File: kpack-image-builder/controllers/kpack_build_controller_test.go (hunk 10 above)
Comment (kieron-dev) [reply kpack-image-builder/controllers/kpack_build_controller_test.go:111]: OK - this was testing finalizer addition, which is now handled by the webhook.

File: controllers/api/v1alpha1/cfapp_webhook_test.go [hunk 12]
```diff
@@ -21,7 +21,7 @@ var _ = Describe("CFAppMutatingWebhook", func() {
 		cfApp = &korifiv1alpha1.CFApp{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      GenerateGUID(),
-				Namespace: namespace,
+				Namespace: rootNamespace,
```
[the remark this answers is not in the document]
Comment (kieron-dev) [reply controllers/api/v1alpha1/cfapp_webhook_test.go]: This was just a rename. There is no actual root/org/space namespace hierarchy set up in these tests. We've reverted the name change.

File: controllers/api/v1alpha1/cfbuild_webhook_test.go [hunk 13]
```diff
@@ -31,7 +31,7 @@ var _ = Describe("CFBuildMutatingWebhook", func() {
 		cfBuild = &korifiv1alpha1.CFBuild{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      cfBuildGUID,
-				Namespace: namespace,
+				Namespace: rootNamespace,
```
[the remark this answers is not in the document]
Comment (kieron-dev) [reply controllers/api/v1alpha1/cfbuild_webhook_test.go]: As above

File: controllers/api/v1alpha1/cfpackage_webhook_test.go [hunk 14]
```diff
@@ -22,7 +22,7 @@ var _ = Describe("CFPackageMutatingWebhook", func() {
 		cfPackage = &korifiv1alpha1.CFPackage{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      GenerateGUID(),
-				Namespace: namespace,
+				Namespace: rootNamespace,
```
[the remark this answers is not in the document]
Comment (kieron-dev) [reply controllers/api/v1alpha1/cfpackage_webhook_test.go]: As above

File: controllers/api/v1alpha1/cfprocess_webhook_test.go (hunk 11 above)
Comment (kieron-dev) [reply controllers/api/v1alpha1/cfprocess_webhook_test.go]: As above

**Review by kieron-dev — APPROVED** [review]

**Comment by kieron-dev** [comment]
@matt-royal - Thanks for the review. We've reverted those changes you highlighted.

Hope you don't mind us dismissing that review now so we can get the change merged.
